File: src/llm/GeminiProvider.ts
/**
 * GeminiProvider - Concrete implementation of BaseLLMProvider using
 * the Google Generative AI (Gemini) SDK.
 *
 * This is the only file in the codebase that imports @google/generative-ai.
 * Swapping to a different provider only requires creating a new implementation
 * of BaseLLMProvider — no other code changes needed.
 *
 * Includes automatic retry with exponential backoff for rate-limit (429) errors.
 */

import { GoogleGenerativeAI, GenerativeModel } from "@google/generative-ai";
import {
  BaseLLMProvider,
  LLMGenerationOptions,
  LLMResponse,
} from "./BaseLLMProvider";

export class GeminiProvider implements BaseLLMProvider {
  readonly providerName = "gemini";
  private readonly model: GenerativeModel;
  private readonly modelName: string;
  private readonly maxRetries: number;

  constructor(
    apiKey: string,
    modelName: string = "gemini-2.0-flash",
    maxRetries: number = 5
  ) {
    if (!apiKey) {
      throw new Error("GEMINI_API_KEY is required to initialize GeminiProvider");
    }
    const genAI = new GoogleGenerativeAI(apiKey);
    this.modelName = modelName;
    this.model = genAI.getGenerativeModel({ model: modelName });
    this.maxRetries = maxRetries;
  }

  async generateResponse(
    prompt: string,
    context?: string,
    options?: LLMGenerationOptions
  ): Promise<LLMResponse> {
    const systemPrompt =
      options?.systemPrompt ||
      "You are a helpful Shopcard assistant. Answer the user's question based ONLY on the provided context. If the context does not contain enough information to answer, say so clearly. Do not make up information.";

    // Build the full prompt with context
    let fullPrompt = `${systemPrompt}\n\n`;

    if (context) {
      fullPrompt += `--- Retrieved Context ---\n${context}\n--- End Context ---\n\n`;
    }

    fullPrompt += `User Question: ${prompt}\n\nAnswer:`;

    return this.executeWithRetry(fullPrompt, options);
  }

  /**
   * Execute the LLM call with automatic retry and exponential backoff
   * for rate-limit (429) errors.
   */
  private async executeWithRetry(
    fullPrompt: string,
    options?: LLMGenerationOptions
  ): Promise<LLMResponse> {
    let lastError: Error | null = null;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      try {
        const result = await this.model.generateContent({
          contents: [{ role: "user", parts: [{ text: fullPrompt }] }],
          generationConfig: {
            maxOutputTokens: options?.maxTokens ?? 1024,
            temperature: options?.temperature ?? 0.3,
          },
        });

        const response = result.response;
        const text = response.text();

        return {
          text,
          provider: this.providerName,
          model: this.modelName,
        };
      } catch (error: any) {
        lastError = error;
        const errorMessage: string = error.message || "";

        // Check if this is a rate-limit error (429)
        if (errorMessage.includes("429") || errorMessage.includes("Too Many Requests") || errorMessage.includes("quota")) {
          if (attempt < this.maxRetries) {
            // Parse server-suggested retry delay, or use exponential backoff
            const serverDelay = this.parseRetryDelay(errorMessage);
            const backoffDelay = Math.pow(2, attempt + 1) * 1000; // 2s, 4s, 8s, 16s, 32s
            const waitMs = serverDelay ? (serverDelay * 1000) + 1000 : backoffDelay;

            console.warn(
              `[GeminiProvider] Rate limited (attempt ${attempt + 1}/${this.maxRetries}). ` +
              `Retrying in ${(waitMs / 1000).toFixed(1)}s...`
            );

            await this.sleep(waitMs);
            continue;
          }

          console.error(
            `[GeminiProvider] Rate limit exceeded after ${this.maxRetries} retries. ` +
            `Your free-tier quota may be exhausted. Check: https://ai.google.dev/gemini-api/docs/rate-limits`
          );
        } else {
          // Non-retryable error — fail immediately
          console.error(`[GeminiProvider] Error generating response:`, errorMessage);
          throw new Error(`LLM generation failed: ${errorMessage}`);
        }
      }
    }

    throw new Error(
      `LLM generation failed after ${this.maxRetries} retries: ${lastError?.message}`
    );
  }

  /**
   * Try to extract the server-suggested retry delay (in seconds) from the error message.
   * Looks for patterns like "Please retry in 39.226700788s" or "retryDelay":"39s"
   */
  private parseRetryDelay(errorMessage: string): number | null {
    // Match "Please retry in <N>s" or "retryDelay":"<N>s"
    const patterns = [
      /retry in ([\d.]+)s/i,
      /retryDelay["\s:]+(\d+)s/i,
    ];

    for (const pattern of patterns) {
      const match = errorMessage.match(pattern);
      if (match) {
        const seconds = parseFloat(match[1]);
        if (!isNaN(seconds) && seconds > 0 && seconds < 120) {
          return Math.ceil(seconds);
        }
      }
    }
    return null;
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: src_python/llm/gemini_provider.py
"""
GeminiProvider - Concrete implementation of BaseLLMProvider using
the Google GenAI SDK (google.genai).

This is the only file in the codebase that imports google.genai.
Swapping to a different provider only requires creating a new implementation
of BaseLLMProvider — no other code changes needed.

Includes automatic retry with exponential backoff for rate-limit (429) errors.
"""

from __future__ import annotations

import asyncio
import math
import re
from typing import Optional

from google import genai
from google.genai import types

from .base_llm_provider import BaseLLMProvider, LLMGenerationOptions, LLMResponse


class GeminiProvider(BaseLLMProvider):
    def __init__(
        self,
        api_key: str,
        model_name: str = "gemini-2.0-flash",
        max_retries: int = 5,
    ):
        if not api_key:
            raise ValueError("GEMINI_API_KEY is required to initialize GeminiProvider")

        self._client = genai.Client(api_key=api_key)
        self._model_name = model_name
        self._max_retries = max_retries

    @property
    def provider_name(self) -> str:
        return "gemini"

    async def generate_response(
        self,
        prompt: str,
        context: Optional[str] = None,
        options: Optional[LLMGenerationOptions] = None,
    ) -> LLMResponse:
        system_prompt = (
            (options.system_prompt if options and options.system_prompt else None)
            or (
                "You are a helpful Shopcard assistant. Answer the user's question "
                "based ONLY on the provided context. If the context does not contain "
                "enough information to answer, say so clearly. Do not make up information."
            )
        )

        # Build the full prompt with context
        full_prompt = f"{system_prompt}\n\n"

        if context:
            full_prompt += f"--- Retrieved Context ---\n{context}\n--- End Context ---\n\n"

        full_prompt += f"User Question: {prompt}\n\nAnswer:"

        return await self._execute_with_retry(full_prompt, options)

    async def _execute_with_retry(
        self,
        full_prompt: str,
        options: Optional[LLMGenerationOptions] = None,
    ) -> LLMResponse:
        """
        Execute the LLM call with automatic retry and exponential backoff
        for rate-limit (429) errors.
        """
        last_error: Optional[Exception] = None

        for attempt in range(self._max_retries + 1):
            try:
                # google.genai's generate_content is synchronous;
                # run it in a thread to keep the event loop free.
                result = await asyncio.to_thread(
                    self._client.models.generate_content,
                    model=self._model_name,
                    contents=full_prompt,
                    config=types.GenerateContentConfig(
                        max_output_tokens=(
                            options.max_tokens if options and options.max_tokens else 1024
                        ),
                        temperature=(
                            options.temperature if options and options.temperature is not None else 0.3
                        ),
                    ),
                )

                text = result.text

                return LLMResponse(
                    text=text,
                    provider=self.provider_name,
                    model=self._model_name,
                )

            except Exception as error:
                last_error = error
                error_message = str(error)

                # Check if this is a rate-limit error (429)
                if "429" in error_message or "Too Many Requests" in error_message or "quota" in error_message:
                    if attempt < self._max_retries:
                        # Parse server-suggested retry delay, or use exponential backoff
                        server_delay = self._parse_retry_delay(error_message)
                        backoff_delay = math.pow(2, attempt + 1)  # 2s, 4s, 8s, 16s, 32s
                        wait_s = (server_delay + 1) if server_delay else backoff_delay

                        print(
                            f"[GeminiProvider] Rate limited (attempt {attempt + 1}/"
                            f"{self._max_retries}). Retrying in {wait_s:.1f}s..."
                        )

                        await asyncio.sleep(wait_s)
                        continue

                    print(
                        f"[GeminiProvider] Rate limit exceeded after {self._max_retries} "
                        f"retries. Your free-tier quota may be exhausted. "
                        f"Check: https://ai.google.dev/gemini-api/docs/rate-limits"
                    )
                else:
                    # Non-retryable error — fail immediately
                    print(f"[GeminiProvider] Error generating response: {error_message}")
                    raise RuntimeError(f"LLM generation failed: {error_message}") from error

        raise RuntimeError(
            f"LLM generation failed after {self._max_retries} retries: {last_error}"
        )

    @staticmethod
    def _parse_retry_delay(error_message: str) -> Optional[float]:
        """
        Try to extract the server-suggested retry delay (in seconds) from the error message.
        Looks for patterns like "Please retry in 39.226700788s" or "retryDelay":"39s"
        """
        patterns = [
            r"retry in ([\d.]+)s",
            r'retryDelay["\s:]+(\d+)s',
        ]

        for pattern in patterns:
            match = re.search(pattern, error_message, re.IGNORECASE)
            if match:
                seconds = float(match.group(1))
                if 0 < seconds < 120:
                    return math.ceil(seconds)
        return None
